stripe/src/main/java/com/stripe/android/view/PaymentAuthWebViewClient.kt

```diff
@@ -0,0 +1,203 @@
+        private set
+
+    internal var hasLoadedBlank: Boolean = false
+
+    override fun onPageFinished(view: WebView, url: String?) {
+        logger.debug("PaymentAuthWebViewClient#onPageFinished() - $url")
+        super.onPageFinished(view, url)
+
+        if (!hasLoadedBlank) {
+            // hide the progress bar here because doing it in `onPageCommitVisible()`
+            // potentially causes a crash
+            hideProgressBar()
+        }
+
+        if (url != null && isCompletionUrl(url)) {
+            onAuthCompleted()
+        }
+    }
+
+    private fun hideProgressBar() {
+        logger.debug("PaymentAuthWebViewClient#hideProgressBar()")
+
+        isPageLoaded.value = true
+    }
+
+    private fun isAuthenticateUrl(url: String) = isAllowedUrl(url, AUTHENTICATE_URLS)
+
+    private fun isCompletionUrl(url: String) = isAllowedUrl(url, COMPLETION_URLS)
+
+    private fun isAllowedUrl(url: String, allowedUrls: Set<String>): Boolean {
+        for (completionUrl in allowedUrls) {
+            if (url.startsWith(completionUrl)) {
+                return true
+            }
+        }
+
+        return false
+    }
+
+    override fun shouldOverrideUrlLoading(view: WebView, urlString: String): Boolean {
+        logger.debug("PaymentAuthWebViewClient#shouldOverrideUrlLoading() - $urlString")
+        val uri = Uri.parse(urlString)
+        updateCompletionUrl(uri)
+
+        return if (isReturnUrl(uri)) {
+            logger.debug("PaymentAuthWebViewClient#shouldOverrideUrlLoading() - handle return URL")
+            onAuthCompleted()
+            true
+        } else if ("intent".equals(uri.scheme, ignoreCase = true)) {
+            openIntentScheme(uri)
+            true
+        } else if (!URLUtil.isNetworkUrl(uri.toString())) {
+            // Non-network URLs are likely deep-links into banking apps. If the deep-link can be
+            // opened via an Intent, start it. Otherwise, stop the authentication attempt.
+            openIntent(Intent(Intent.ACTION_VIEW, uri))
+            true
+        } else {
+            super.shouldOverrideUrlLoading(view, urlString)
+        }
+    }
+
+    private fun openIntentScheme(uri: Uri) {
+        logger.debug("PaymentAuthWebViewClient#openIntentScheme()")
+        runCatching {
+            openIntent(Intent.parseUri(uri.toString(), Intent.URI_INTENT_SCHEME))
+        }.onFailure {
+            logger.error("Failed to start Intent.", it)
+            onAuthCompleted()
+        }
+    }
+
+    /**
+     * See https://developer.android.com/training/basics/intents/package-visibility-use-cases
+     * for more details on app-to-app interaction.
+     */
+    private fun openIntent(intent: Intent) {
+        logger.debug("PaymentAuthWebViewClient#openIntent()")
+
+        runCatching {
+            activityStarter(intent)
+        }.onFailure {
+            logger.error("Failed to start Intent.", it)
+
+            if (intent.scheme != "alipays") {
```

Comment:
Probably not related to this PR, but seems like this scheme should be a companion object or a constant somewhere?

Reply:
I can add this in a future PR.